gateway-js/src/loadCsdlFromStorage.ts

```diff
@@ -0,0 +1,133 @@
+import { fetch } from 'apollo-server-env';
+import { GraphQLError } from 'graphql';
+
+export const CSDL_QUERY = `#graphql
+  query Csdl($apiKey: String!, $ref: String!) {
+    routerConfig(
+      ref: $ref
+      apiKey: $apiKey
+    ) {
+      __typename
+      ... on RouterConfigResult {
+        id
+        csdl
+      }
+      ... on FetchError {
+        code
+        message
+      }
+    }
+  }
+`;
+
+type CsdlQueryResult = CsdlQuerySuccess | CsdlQueryFailure;
+
+interface CsdlQuerySuccess {
+  data: { routerConfig: RouterConfigResponse };
+}
+
+interface CsdlQueryFailure {
+  data?: { routerConfig: RouterConfigResponse };
+  errors: GraphQLError[];
+}
+
+type RouterConfigResponse = RouterConfigResult | FetchError;
+
+interface RouterConfigResult {
+  __typename: 'RouterConfigResult';
+  id: string;
+  csdl: string;
+  messages: Message[];
+}
+
+interface Message {
+  __typename: 'Message';
+  body: string;
+  level: MessageLevel;
+}
+
+enum MessageLevel {
+  ERROR = 'ERROR',
+  WARN = 'WARN',
+  INFO = 'INFO',
+}
+
+interface FetchError {
+  __typename: 'FetchError';
+  code: FetchErrorCode;
+  message: string;
+}
+
+enum FetchErrorCode {
+  AUTHENTICATION_FAILED = 'AUTHENTICATION_FAILED',
+  ACCESS_DENIED = 'ACCESS_DENIED',
+  UNKNOWN_REF = 'UNKNOWN_REF',
+  RETRY_LATER = 'RETRY_LATER',
```

Comment:
Should we handle this retry later specially?

Reply:
Yes, though it sounds like Gravity has some changes in the works and I'm unsure what these errors are going to look like. I'm going to hold off for now and see if this concept even still exists in the next iteration.